File: projectchess2/PieceFactory.java
package projectchess2;

import projectchess2.pieces.Bishop;
import projectchess2.pieces.King;
import projectchess2.pieces.Knight;
import projectchess2.pieces.Pawn;
import projectchess2.pieces.Queen;
import projectchess2.pieces.Rook;

public class PieceFactory {
	
	private Board board;

	public PieceFactory(Board board) {
		this.board = board;
	}

	public Piece createPiece(Pos pos, PieceType type, Side side) {
		Piece result = null;
		
		switch(type){
			case EMPTY:
				result = new EmptyPiece();
				break;
			case BISHOP:
				result = new Bishop(board, pos, side);
				break;
			case KING:
				result = new King(board, pos, side);
				break;
			case KNIGHT:
				result = new Knight(board, pos, side);
				break;
			case PAWN:
				result = new Pawn(board, pos, side);
				break;
			case QUEEN:
				result = new Queen(board, pos, side);
				break;
			case ROOK:
				result = new Rook(board, pos, side);
				break;
			default: break;
		}
		
		return result;
	}

	public Piece createPieceByTransposition(Pos dest, Piece piece) {
		Piece newPiece = createPiece(dest, piece.getType(), piece.getSide());
		if(newPiece instanceof Pawn){
			((Pawn) newPiece).setMoved(true);
		}
		return newPiece;
	}
	
	public Piece copyPiece(Piece toCopy){
		Piece newPiece = createPiece(toCopy.getPos(), toCopy.getType(), toCopy.getSide());
		if(toCopy instanceof Pawn){
			((Pawn) newPiece).setMoved(((Pawn) toCopy).hasMoved());
		}
		return newPiece;
		
	}

}

File: typescript/ProjectChess/src/model/PieceFactory.ts
class PieceFactory{

    static createPiece(board: BoardModel, pos: Pos, color: Color, type: PieceType): PieceModel{
        var newPiece: PieceModel;
        switch(type){
            case PieceType.ROOK: newPiece = new RookModel(board, pos, color); break;
            case PieceType.BISHOP: newPiece = new BishopModel(board, pos, color); break;
            case PieceType.PAWN: newPiece = new PawnModel(board, pos, color); break;
            case PieceType.KING: newPiece = new KingModel(board, pos, color); break;
            case PieceType.KNIGHT: newPiece = new KnightModel(board, pos, color); break;
            case PieceType.QUEEN: newPiece = new QueenModel(board, pos, color); break;
            case PieceType.GENERAL: newPiece = new GeneralModel(board, pos, color); break;
            case PieceType.MINISTER: newPiece = new MinisterModel(board, pos, color); break;
            case PieceType.GIRAFFE_RIDER: newPiece = new GiraffeRiderModel(board, pos, color); break;
            case PieceType.WAR_MACHINE: newPiece = new WarMachineModel(board, pos, color); break;
            case PieceType.CAMEL_RIDER: newPiece = new CamelRiderModel(board, pos, color); break;
            case PieceType.ELEPHANT_RIDER: newPiece = new ElephantRiderModel(board, pos, color); break;
            case PieceType.PICKET: newPiece = new PicketModel(board, pos, color); break;
        }
        return newPiece;
    }

    static createPieceByTransposition(pos: Pos, piece: PieceModel): PieceModel{
        var newPiece: PieceModel = PieceFactory.createPiece(piece.getBoardModel(), pos, piece.getColor(), piece.getType());
        piece.giveInternalAttributes(newPiece);
        return newPiece;
    }
}